File: app-tracker-automation/pages/base_page.py
"""
Base Page class for Page Object Model pattern.
Provides common functionality for all page objects.
"""

from playwright.sync_api import Page, expect
from utils.logger import Logger
from utils.waits import Waits


class BasePage:
    """Base class for all page objects"""
    
    def __init__(self, page: Page):
        self.page = page
        self.logger = Logger()
        self.waits = Waits(page)
    
    def navigate_to(self, url: str):
        """Navigate to specified URL"""
        self.logger.info(f"Navigating to: {url}")
        self.page.goto(url)
    
    def get_title(self) -> str:
        """Get page title"""
        return self.page.title()
    
    def wait_for_page_load(self, timeout: int = 30000):
        """Wait for page to fully load"""
        self.waits.wait_for_page_load(timeout)
    
    def take_screenshot(self, name: str):
        """Take screenshot"""
        self.page.screenshot(path=f"screenshots/{name}.png")
    
    def is_element_visible(self, locator: str) -> bool:
        """Check if element is visible"""
        return self.page.locator(locator).is_visible()
    
    def click_element(self, locator: str):
        """Click on element"""
        self.page.locator(locator).click()
    
    def fill_text(self, locator: str, text: str):
        """Fill text in input field"""
        self.page.locator(locator).fill(text)

File: ts-framework/pages/base_page.ts
import { Page } from '@playwright/test';
import { Logger } from '../utils/logger';
import { Waits } from '../utils/waits';

export class BasePage {
  protected page: Page;
  protected logger: Logger;
  protected waits: Waits;

  constructor(page: Page) {
    this.page = page;
    this.logger = new Logger('BasePage');
    this.waits = new Waits(page);
  }

  async navigateTo(url: string): Promise<void> {
    this.logger.info(`Navigating to: ${url}`);
    await this.page.goto(url);
  }

  async getTitle(): Promise<string> {
    return await this.page.title();
  }

  async waitForPageLoad(timeout: number = 30000): Promise<void> {
    await this.waits.waitForPageLoad(timeout);
  }

  async takeScreenshot(name: string): Promise<void> {
    await this.page.screenshot({ path: `screenshots/${name}.png` });
  }

  async isElementVisible(locator: string): Promise<boolean> {
    return await this.page.locator(locator).isVisible();
  }

  async clickElement(locator: string): Promise<void> {
    await this.page.locator(locator).click();
  }

  async fillText(locator: string, text: string): Promise<void> {
    await this.page.locator(locator).fill(text);
  }
}